sklearn/base.py

```diff
@@ -324,6 +326,30 @@ def _get_tags(self):
         tags.update(collected_tags)
         return tags
 
+    def _validate_n_features(self, X, check_n_features):
+        if check_n_features:
+            if not hasattr(self, 'n_features_in_'):
+                raise RuntimeError(
+                    "check_n_features is True but there is no n_features_in_ "
+                    "attribute."
+                )
+            if X.shape[1] != self.n_features_in_:
+                raise ValueError(
+                    'X has {} features, but this {} is expecting {} features '
+                    'as input.'.format(X.shape[1], self.__class__.__name__,
+                                       self.n_features_in_)
+                )
+        self.n_features_in_ = X.shape[1]
+
+    def validate_X(self, X, check_n_features=False, **check_array_params):
+        X = check_array(X, **check_array_params)
+        self._validate_n_features(X, check_n_features)
+        return X
+
+    def validate_X_y(self, X, y, check_n_features=False, **check_X_y_params):
```

Comment:
I don't think it makes sense to have this, certainly not public, on unsupervised estimators.